WIP business rules

elainekamlley opened the pull request:
This is the first draft of SBA business rules and validations and will be updated.

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+---
+As an SBA user, I need to see the business rules being applied to my source data
+
+Business rules or validation rules will be applied to ensure that submitted data by the US Small Business Administration meets the DATA Act criteria. Goals of these validation rules will improve cleanliness, accuracy, and meaningfulness of the data set. Below is a draft of proposed rules and are subject to change. 
```
It might be helpful to be specific about when the rules will be applied. For example, "agency submissions will be checked against these rules before the data submission is accepted by Treasury" or something like that.

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+*Standard validation*
+
+Required fields: 
+	Required field validation ensures that users input data in required fields. If field is empty, an error will be raised. Required fields to be determined in the [DATA act schema](https://github.com/18F/intercessor/blob/master/schema/data-act-schema.png)
```
* Just to be as specific as possible: "If a _required_ field is empty, and error will be raised." 
* since the DATA Act schema doesn't really determine the required fields (it just reflects them), might be more accurate to say that "the DATA Act schema documents the required fields"

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+	Data type validation ensures that the data types match with the [DATA Act schema](https://github.com/18F/intercessor/blob/master/schema/data-act-schema.png) . For example, `parentAward` data type required is a string, If a number is entered an error will be raised. 
+
+
+*Awardee validation:*  
```
Good idea, having a separate section for awardee validation. What do you think about breaking out the individual validations as a list so they're easier to read?

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+		Address validation ensures that users input data in the Awardee Address fields. If fields are empty, an error will be raised. 
+		state and zip codes validation will ensure that Awardee Zip/Postal Code is valid by looking up the first five characters of the value in a zip code object that contains a record for every valid zip code in the US. If the zip code is not found in the zip code object, an error will be raised. 
+	
+Award validation: 
```
I think we could flesh this out a bit by saying something like: 

The following fields will be validated against their respective authoritative sources:
* Object Class 
* CFDA Code
* NAICS Code

*at some point, we'll have to document what we'll use for the authoritative sources

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+Questions: 
+
+- How do you keep track of changing addresses? 
+- Is it more important that the validation be associated with the current address or does it need to be when the award was issued? 
```
This is important question. It also applies more broadly, so you might add a question like "In general, how do we want to track data that changes over time?"

bsweger commented:
Thanks for this great start, @elainekamlley. Added some notes inline!

micahsaul commented on sba-business-rules.md:
```diff
@@ -0,0 +1,37 @@
+- How do you keep track of changing addresses? 
+- Is it more important that the validation be associated with the current address or does it need to be when the award was issued? 
+- Will you have a list of valid state codes and zip codes? 
+- What are the authoritative sources to validate addresses, NACIS code, CFDA code? 
```
For addresses, I assume SAM is the authoritative source.

But yes, important question about how we track data over time. My gut feeling is that, since Action seems to be the agreed upon most granular entity, each record should have the current information _**at action time**_. But I'm not committed to that.

bsweger commented:
@micahsaul just curious--in your adventures with the SAM API, did you come across any info about whether or not you can query vendor info at a point in time?

elainekamlley commented:
Hey All, 

Thanks for such great feedback @bsweger @micahsaul . I took another swing at it and added all of the notes and cleaned up the formatting. Please keep the feedback coming, can make additional edits this afternoon.

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,62 @@
+- Required field validation ensures that users input data in required fields. If required field is empty, an error will be raised. [DATA act schema](https://github.com/18F/intercessor/blob/master/schema/data-act-schema.png) documents the required fields.
+
+*Data types:*
+- Data type validation ensures that the data types match with the [DATA Act schema](https://github.com/18F/intercessor/blob/master/schema/data-act-schema.png) . For example, `parentAward` data type required is a string, If a number is entered an error will be raised. 
```
is a string**.** If a number is entered (replaced _,_ with _._).

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,62 @@
+**Awardee validation:**  
+
+*Award id:*
+- Award id validation ensures that `Award id` is an exact number length (if it is not blank). This length can be determined and set. Additionally, award id validation ensures DUNSNumber and SAM number match. If either of those criterias are not met an error will be raised
```
will be raised. (added _._)

bsweger commented on sba-business-rules.md:
```diff
@@ -0,0 +1,62 @@
+*Address:* 
+
+- Address validation ensures that users input data in the Awardee Address fields. If address fields are empty, an error will be raised. 
+- state and zip codes validation will ensure that Awardee Zip/Postal Code is valid by looking up the first five characters of the value in a zip code object that contains a record for every valid zip code in the US. If the zip code is not found in the zip code object, an error will be raised. 
```
Capitalize _state_ for consistency w/ other bullets?

bsweger commented:
Thanks @elainekamlley. This looks great--I just added a few tiny line notes related to consistency--nothing substantial. Thanks so much for putting this together.
